Kernel/Arch/PageFault.h

```diff
@@ -44,21 +57,42 @@ class PageFault {
     };
 
     VirtualAddress vaddr() const { return m_vaddr; }
-    u16 code() const { return m_code; }
+    u16 code() const
+    {
+        u16 code = 0;
+        code |= (u16)m_type;
+        code |= (u16)m_access;
+        code |= m_execution_mode == ExecutionMode::User ? PageFaultFlags::UserMode : 0;
+        code |= m_is_reserved_bit_violation ? PageFaultFlags::ReservedBitViolation : 0;
+        code |= m_is_instruction_fetch ? PageFaultFlags::InstructionFetch : 0;
+        return code;
+    }
+
+    void set_type(Type type) { m_type = type; }
+    Type type() const { return m_type; }
 
-    Type type() const { return (Type)(m_code & 1); }
-    Access access() const { return (Access)(m_code & 2); }
+    void set_access(Access access) { m_access = access; }
+    Access access() const { return m_access; }
 
-    bool is_not_present() const { return (m_code & 1) == PageFaultFlags::NotPresent; }
-    bool is_protection_violation() const { return (m_code & 1) == PageFaultFlags::ProtectionViolation; }
-    bool is_read() const { return (m_code & 2) == PageFaultFlags::Read; }
-    bool is_write() const { return (m_code & 2) == PageFaultFlags::Write; }
-    bool is_user() const { return (m_code & 4) == PageFaultFlags::UserMode; }
-    bool is_supervisor() const { return (m_code & 4) == PageFaultFlags::SupervisorMode; }
-    bool is_instruction_fetch() const { return (m_code & 16) == PageFaultFlags::InstructionFetch; }
+    void set_mode(ExecutionMode execution_mode) { m_execution_mode = execution_mode; }
+    ExecutionMode mode() const { return m_execution_mode; }
+
+    bool is_not_present() const { return m_type == Type::PageNotPresent; }
+    bool is_protection_violation() const { return m_type == Type::ProtectionViolation; }
+    bool is_read() const { return m_access == Access::Read; }
+    bool is_write() const { return m_access == Access::Write; }
+    bool is_user() const { return m_execution_mode == ExecutionMode::User; }
+    bool is_supervisor() const { return m_execution_mode == ExecutionMode::Kernel; }
```

Comment:
given the internals, `is_kernel` would sound more natural here?

Reply:
Yes, that would be better indeed, changed!